app/src/main/java/com/github/sdpcoachme/database/FireDatabase.kt

```diff
@@ -49,12 +53,55 @@ class FireDatabase(databaseReference: DatabaseReference) : Database {
         }
     }
 
+    override fun addGroupEvent(groupEvent: GroupEvent): CompletableFuture<Void> {
+        var errorPreventionFuture = CompletableFuture<Void>()
+
+        if (groupEvent.participants.size > groupEvent.maxParticipants) {
+            errorPreventionFuture.completeExceptionally(Exception("Group event should not be full, initially"))
+        } else if (groupEvent.participants.isEmpty()) {
+            errorPreventionFuture.completeExceptionally(Exception("Group event must have at least 2 participants"))
+        } else if (LocalDateTime.parse(groupEvent.event.start).isBefore(LocalDateTime.now())) {
+            errorPreventionFuture.completeExceptionally(Exception("Group event cannot be in the past"))
+        } else {
+            errorPreventionFuture = setChild(groupEvents, groupEvent.groupEventId, groupEvent).thenCompose {
+                registerForGroupEvent(groupEvent.organiser, groupEvent.groupEventId)
+            }
+        }
+
+        return errorPreventionFuture
+    }
+
+    override fun registerForGroupEvent(email: String, groupEventId: String): CompletableFuture<Void> {
+        val id = email.replace('.', ',')
+        return getGroupEvent(groupEventId).thenCompose { groupEvent ->
+            val hasCapacity = groupEvent.participants.size < groupEvent.maxParticipants
+            if (!hasCapacity) {
+                val failingFuture = CompletableFuture<Void>()
+                failingFuture.completeExceptionally(Exception("Group event is full"))
+                failingFuture
+            } else {
+                val updatedGroupEvent = groupEvent.copy(participants = groupEvent.participants + email)
+                setChild(groupEvents, groupEventId, updatedGroupEvent).thenCompose {
+                    getSchedule(email, EventOps.getStartMonday()).thenCompose { s ->
+                        val updatedSchedule = s.copy(groupEvents = s.groupEvents + groupEventId)
+                        setChild(schedule, id, updatedSchedule) // Return updated schedule?
+                    }
+                }
+            }
+        }
+    }
+
     override fun getSchedule(email: String, currentWeekMonday: LocalDate): CompletableFuture<Schedule> {
         val id = email.replace('.', ',')
         return getChild(schedule, id).thenApply { it.getValue(Schedule::class.java)!! }
             .exceptionally { Schedule() }
     }
 
+    override fun getGroupEvent(groupEventId: String): CompletableFuture<GroupEvent> {
```

Comment:
I believe, it would be best to also call replace on the groupEventId to replace "."'s by ","'s